packages/data-driver-postgres/src/index.ts

```diff
@@ -33,35 +38,71 @@ export default class DataDriverPostgres implements DataDriver {
 		await this.#pool.end();
 	}
 
-	async getDataFromSource(pool: pg.Pool, sql: ParameterizedSqlStatement): Promise<{ poolClient: any; stream: any }> {
-		const poolClient: PoolClient = await pool.connect();
-		const queryStream = new QueryStream(sql.statement, sql.parameters);
-		const stream = poolClient.query(queryStream);
-		stream.on('end', () => poolClient?.release());
-		const webStream = Readable.toWeb(stream);
-
-		return {
-			poolClient,
-			stream: webStream,
-		};
+	/**
+	 * Opens a stream for the given SQL statement.
+	 *
+	 * @param pool the PostgreSQL client pool
+	 * @param sql A parameterized SQL statement
+	 * @returns A readable web stream for the query results
+	 * @throw An error when the query cannot be performed
+	 */
+	async getDataFromSource(
+		pool: pg.Pool,
+		sql: ParameterizedSqlStatement
+	): Promise<ReadableStream<Record<string, unknown>>> | never {
+		try {
+			const poolClient = await pool.connect();
+			const queryStream = new QueryStream(sql.statement, sql.parameters);
+			const stream = poolClient.query(queryStream);
+
+			stream.on('end', () => poolClient.release());
+			stream.on('error', () => poolClient.release());
+
+			return Readable.toWeb(stream);
+		} catch (error: any) {
+			throw new Error('Failed to query the database: ', error);
+		}
 	}
 
-	async query(query: AbstractQuery): Promise<ReadableStream> {
-		let client: PoolClient | null = null;
+	/**
+	 * Converts the abstract query into PostgreSQL and executes it.
+	 *
+	 * @param abstractSql The abstract query
+	 * @returns The database results converted to a nested object
+	 * @throws An error when the conversion or the database request fails
+	 */
+	private async queryDatabase(abstractSql: AbstractSqlQuery): Promise<ReadableStream<Record<string, unknown>>> {
+		let statement;
+		let parameters;
+
+		try {
+			statement = convertToActualStatement(abstractSql.clauses);
+			parameters = convertParameters(abstractSql.parameters);
+		} catch (error: any) {
+			throw new Error('Failed to convert the query the database: ', error);
+		}
+
+		const stream = await this.getDataFromSource(this.#pool, { statement, parameters });
```

Comment:
We should probably let errors from our own functions bubble up instead of rethrowing them. Also, this effectively types `statement` and `parameters` as `any`.

Reply:
We still bubble the errors form the conversion functions up, but with this try catch in place the consumer will get a more detailed stack trace. so now the consumer will receive an error saying that the conversion went wrong first and then sees the exact error from the converting step. that was my intention

Reply:
for the expander, I removed the additional throwing of an error and enhance the message of the inner errors a bit instead

Reply:
now with the changes, we also have the correct types again for `statement` and `parameter`

Reply:
I feel like we should handle this by defining separate error classes like `QueryConvertError` etc. which add the needed context to error messages instead of re-throwing. All the try-catches make the code a lot less readable.

Reply:
I like the idea! but let's better do it later on. readability of those couple of lines should not be our main concern atm since we're already behind schedule